firebase-firestore/src/main/java/com/google/firebase/firestore/model/InternalDocument.java

```diff
@@ -0,0 +1,33 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.model;
+
+/** Interface exposing underlying document information that is not visible in {@code Document}. */
+public interface InternalDocument extends Document {
```

Comment:
I don't quite understand the need for this indirection. Can you explain your reasoning?

Reply:
I guess I wanted a bit type safety, that `getVersion` and `getReadTime` can only be accessed via this interface, instead of setting them to 0. And `DocumentSnapshot` holds a `Document` so no access to version/readtime to user code.

I feel safer this way, comparing to reviewing all places where they are used and reason whether 0 makes sense there. Setting them to 0 also felt "wrong" because we know they are at least the versions they are set now, and setting to 0 gives up that information.

Reply:
My concern here is the same as with the other change: It adds a layer of indirection which adds porting work and code size. We already use SnapshotVersion.NONE as the "non-existing" snapshot version in a lot of places in our codebase.

If you want to achieve what you are after, you could make the missing versions nullable - but I really thing we should just continue to use SnapshotVersion.NONE.

I also do not quite understand where the name "Internal" comes from. I think you mean something like "FoundDocument", which we used to have but moved away from.

Reply:
Done.